force_bdss/data_sources/base_data_source_model.py

```diff
@@ -46,43 +46,218 @@ class BaseDataSourceModel(BaseModel):
+
+    # -------------------
+    #      Defaults
+    # -------------------
+
+    def _input_slot_info_default(self):
+        """Default list of InputSlotInfo object, based on length and
+        attributes of Slots tuple returned by slots method on associated
+        BaseDataSource"""
+
+        input_slots, _ = self._generate_slots()
+
+        return [
+            InputSlotInfo(
+                name='', type=slot.type,
+                description=slot.description
+            )
+            for slot in input_slots
+        ]
+
+    def _output_slot_info_default(self):
+        """Default list of OutputSlotInfo object, based on length and
+        attributes of Slots tuple returned by slots method on associated
+        BaseDataSource"""
+        _, output_slots = self._generate_slots()
+
+        return [
+            OutputSlotInfo(
+                name='', type=slot.type,
+                description=slot.description
+            )
+            for slot in output_slots
+        ]
+
+    # -------------------
+    #      Listeners
+    # -------------------
+
+    @on_trait_change("+changes_slots")
+    def _trigger_changes_slots(self, obj, name, new):
+        changes_slots = self.traits()[name].changes_slots
+
+        if changes_slots:
+            self._reset_data_source_slots()
+            self.changes_slots = True
+
+    # -------------------
+    #  Protected Methods
+    # -------------------
+
+    def __getstate__(self):
+        state = pop_dunder_recursive(super().__getstate__())
+        state["input_slot_info"] = [
+            x.__getstate__() for x in self.input_slot_info
+        ]
+        state["output_slot_info"] = [
+            x.__getstate__() for x in self.output_slot_info
+        ]
+
+        return state
+
+    # -------------------
+    #   Private Methods
+    # -------------------
+
+    def _generate_slots(self):
+        """Returns slots generated by the DataSource associated with the
+        assigned factory trait"""
+
+        data_source = self.factory.create_data_source()
 
         try:
             input_slots, output_slots = data_source.slots(self)
         except Exception:
             logger.exception(
                 "Unable to retrieve slot information from data source"
-                " created by factory '%s', plugin '%s'. This might "
+                " created by factory {}, plugin {}. This might "
                 "indicate a programming error.",
                 self.factory.id,
                 self.factory.plugin.id
             )
             raise
 
+        return input_slots, output_slots
+
+    def _reset_data_source_slots(self):
+        """Reset input_slot_info and output_slot_info attributes to
+        their defaults, based on the return value of a BaseDataSource
+        slots method"""
+
+        self.input_slot_info = self._input_slot_info_default()
+        self.output_slot_info = self._output_slot_info_default()
+
+    def _sync_slot_info(self, new_info, old_info):
```

Comment:
The method provides a way to sync the `name` attribute between two `InputSlotInfo` or `OutputSlotInfo` objects, with some extra type checking.